firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,508 @@
+        bint variable
+
+    variable = mesh.variable_layers
+    cell_closures = mesh.cell_closure
+    if variable:
+        layer_extents = mesh.layer_extents
+        if offset is None:
+            raise ValueError("Offset cannot be None with variable layer extents")
+
+    nclosure = cell_closures.shape[1]
+
+    # Extract ordering from FInAT element entity DoFs
+    ndofs_list = []
+    flat_index_list = []
+
+    for dim in sorted(entity_dofs.keys()):
+        for entity_num in xrange(len(entity_dofs[dim])):
+            dofs = entity_dofs[dim][entity_num]
+
+            ndofs_list.append(len(dofs))
+            flat_index_list.extend(dofs)
+
+    # Coerce lists into C arrays
+    assert nclosure == len(ndofs_list)
+    dofs_per_cell = len(flat_index_list)
+
+    CHKERR(PetscMalloc1(nclosure, &ceil_ndofs))
+    CHKERR(PetscMalloc1(dofs_per_cell, &flat_index))
+
+    for i in range(nclosure):
+        ceil_ndofs[i] = ndofs_list[i]
+    for i in range(dofs_per_cell):
+        flat_index[i] = flat_index_list[i]
+
+    # Fill cell nodes
+    cStart = 0
+    cEnd = mesh._swarm.getLocalSize()
+    cell_nodes = np.empty((cEnd - cStart, dofs_per_cell), dtype=IntType)
+    cell_numbering = mesh._cell_numbering
+    for c in range(cStart, cEnd):
+        k = 0
+        CHKERR(PetscSectionGetOffset(cell_numbering.sec, c, &cell))
+        for i in range(nclosure):
+            entity = cell_closures[cell, i]
+            CHKERR(PetscSectionGetDof(global_numbering.sec, entity, &ndofs))
+            if ndofs > 0:
+                CHKERR(PetscSectionGetOffset(global_numbering.sec, entity, &off))
+                # The cell we're looking at the entity through is
+                # higher than the lowest cell the column touches, so
+                # we need to offset by the difference from the bottom.
+                if variable:
+                    off += offset[flat_index[k]]*(layer_extents[c, 0] - layer_extents[entity, 0])
+                for j in range(ceil_ndofs[i]):
+                    cell_nodes[cell, flat_index[k]] = off + j
+                    k += 1
+
+    CHKERR(PetscFree(ceil_ndofs))
+    CHKERR(PetscFree(flat_index))
+    return cell_nodes
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def reordered_coords(PETSc.DM swarm, PETSc.Section global_numbering, shape):
+    """Return coordinates for the swarm, reordered according to the
+    global numbering permutation for the coordinate function space.
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    Shape is a tuple of (mesh.num_vertices(), geometric_dim)."""
+    cdef:
+        PetscInt v, vStart, vEnd, offset
+        PetscInt i, dim = shape[1]
+        np.ndarray[PetscReal, ndim=2, mode="c"] swarm_coords, coords
+
+    # get coords field - NOTE it isn't copied so could have GC issues!
+    swarm_coords = swarm.getField("DMSwarmPIC_coor").reshape(shape)
```

Comment:
What does this comment mean?

Reply:
I was worried that this might cause problems with the garbage collector if something goes wrong given that you have to explicitly get and restore swarm fields. Unlikely to be a real worry though

Reply:
updated comment